### PR TITLE
MueLu: Fixes

### PR DESCRIPTION
**Pull request by cgcgcg**
@trilinos/muelu 

## Motivation
- Fix a bad view in `GetMatrixDiagonal`. #12767 
- Silence some warnings.
- Fix build errors in research code.

### CODE REVIEW

File: packages/muelu/src/Utils/MueLu_UtilitiesBase_def.hpp [hunk 1]
```diff
@@ -1012,6 +1012,9 @@ DetectDirichletRows_kokkos(const Xpetra::Matrix<SC, LO, GO, NO>& A,
     Kokkos::deep_copy(boundaryNodes2, boundaryNodes);
     return boundaryNodes2;
   }
+  // CAG: No idea why this is needed to avoid "warning: missing return statement at end of non-void function"
```
Comment (GrahamBenHarper) [inline packages/muelu/src/Utils/MueLu_UtilitiesBase_def.hpp:1015]: I'm guessing this is something about the `if constexpr` not being handled properly by the compiler? is this compiler specific, or does it impact many compilers?

File: packages/muelu/research/graham/mf_example_01.cpp [hunk 2]
```diff
@@ -268,10 +268,10 @@ class TridiagonalOperator : public Xpetra::Operator<Scalar, LocalOrdinal, Global
   }
 
   //! Returns the Xpetra::Map object associated with the domain of this operator.
-  Teuchos::RCP<const map_type> getDomainMap() const { return opMap_; }
+  const Teuchos::RCP<const map_type> getDomainMap() const { return opMap_; }
```
Comment (GrahamBenHarper) [inline packages/muelu/research/graham/mf_example_01.cpp:271]: good catch; thanks!

File: packages/muelu/src/Utils/MueLu_UtilitiesBase_def.hpp (hunk 1 above)
Comment (cgcgcg) [reply packages/muelu/src/Utils/MueLu_UtilitiesBase_def.hpp:1015]: I have only seen this on cuda builds.

**Review by GrahamBenHarper — APPROVED** [review]

File: packages/muelu/src/Utils/MueLu_UtilitiesBase_def.hpp [hunk 3]
```diff
@@ -200,8 +200,8 @@ UtilitiesBase<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
   const CrsMatrixWrap* crsOp = dynamic_cast<const CrsMatrixWrap*>(&A);
   if ((crsOp != NULL) && (rowMap->lib() == Xpetra::UseTpetra)) {
     using local_vector_type = typename Vector::dual_view_type::t_dev_um;
-    using execution_space   = typename local_vector_type::execution_space;
-    Kokkos::View<size_t*, execution_space> offsets("offsets", rowMap->getLocalNumElements());
+    using device_type       = typename CrsGraph::device_type;
+    Kokkos::View<size_t*, device_type> offsets("offsets", rowMap->getLocalNumElements());
```
[the remark this answers is not in the document]
Comment (cgcgcg) [reply packages/muelu/src/Utils/MueLu_UtilitiesBase_def.hpp:204]: It only errored out in a UVM build (which is not covered by the AT).

File: packages/muelu/src/Utils/MueLu_UtilitiesBase_def.hpp (hunk 1 above)
Comment (cgcgcg) [reply packages/muelu/src/Utils/MueLu_UtilitiesBase_def.hpp:1015]: Yep, that's the one I tried first but it didn't get rid of the warning.

**Pull request merged**